ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java

```diff
@@ -749,7 +839,140 @@ private void doContentAddAndReadTest(String digestAlgorithm, ByteBuffer content,
+   * {@link NettyRequest#readInto(AsyncWritableChannel, Callback)} call.
+   * @throws Exception
+   */
+  private void doBackPressureTest(String digestAlgorithm, ByteBuffer content, List<HttpContent> httpContents,
+      int numChunksToAddBeforeRead)
+      throws Exception {
+    if (numChunksToAddBeforeRead < 0 || numChunksToAddBeforeRead > httpContents.size()) {
+      throw new IllegalArgumentException("Illegal value of numChunksToAddBeforeRead");
+    }
+
+    MockChannel channel = new MockChannel();
+    int expectedMaxMessagesPerRead = channel.config().getMaxMessagesPerRead();
+    final NettyRequest nettyRequest = createNettyRequest(HttpMethod.POST, "/", null, channel);
+    byte[] wholeDigest = null;
+    if (digestAlgorithm != null && !digestAlgorithm.isEmpty()) {
+      MessageDigest digest = MessageDigest.getInstance(digestAlgorithm);
+      digest.update(content);
+      wholeDigest = digest.digest();
+      content.rewind();
+      nettyRequest.setDigestAlgorithm(digestAlgorithm);
+    }
+
+    final AtomicInteger queuedReads = new AtomicInteger(0);
+    ByteBufferAsyncWritableChannel writeChannel = new ByteBufferAsyncWritableChannel();
+    ReadIntoCallback callback = new ReadIntoCallback();
+
+    channel.setChannelReadCallback(new MockChannel.ChannelReadCallback() {
+      @Override
+      public void onRead() {
+        queuedReads.incrementAndGet();
+      }
+    });
+
+    int addedCount = 0;
+    Future<Long> future = null;
+    boolean suspended = false;
+    int bytesToVerify = 0;
+    while (addedCount < httpContents.size()) {
+      if (suspended) {
+        assertEquals("There should have been no reads queued when over buffer watermark", 0, queuedReads.get());
+        if (future == null) {
+          future = nettyRequest.readInto(writeChannel, callback);
+        }
+        int chunksRead = readAndVerify(bytesToVerify, writeChannel, content);
+        assertEquals("Number of reads triggered is not as expected", chunksRead, queuedReads.get());
+        // collapse many reads into one
+        queuedReads.set(1);
+        bytesToVerify = 0;
+        suspended = false;
+      } else {
+        assertEquals("There should have been only one read queued", 1, queuedReads.get());
+        queuedReads.set(0);
+        if (future == null && addedCount == numChunksToAddBeforeRead) {
+          future = nettyRequest.readInto(writeChannel, callback);
+        }
+        final HttpContent httpContent = httpContents.get(addedCount);
+        bytesToVerify += (httpContent.content().readableBytes());
+        suspended = bytesToVerify >= NettyRequest.bufferWatermark;
+        addedCount++;
+        nettyRequest.addContent(httpContent);
+        int expectedRefCountOnAdd = future == null || httpContent.content().nioBufferCount() > 0 ? 2 : 1;
+        assertEquals("Reference count is not as expected", expectedRefCountOnAdd, httpContent.refCnt());
+      }
+    }
```

Comment:
It seems like you could still cover the case where content gets added regardless of being over the watermark  with yet another flag passed in as an argument - but I realize it can make these checks complicated and may even be outside the scope of what you want to test here.